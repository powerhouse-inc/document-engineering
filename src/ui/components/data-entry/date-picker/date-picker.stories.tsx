import type { Meta, StoryObj } from '@storybook/react'
import { withTimestampsAsISOStrings } from '../../../../scalars/index.js'
import { getDefaultArgTypes, StorybookControlCategory } from '../../../../scalars/lib/storybook-arg-types.js'
import { FORMAT_MAPPING } from '../date-time-picker/utils.js'
import { DatePicker } from './date-picker.js'

/**
 * The `DatePicker` component provides an input field for selecting dates.
 * It supports multiple configuration properties like:
 * - label
 * - description
 * - minDate
 * - maxDate
 * - dateFormat
 * - weekStart
 * - autoClose
 *
 * Features include:
 * - Customizable date format
 * - Min/Max date restrictions
 * - Past/Future date restrictions
 * - Configurable week start day
 * - Auto-close functionality
 * - Custom placeholder support
 *
 * ## Calendar Customization
 * The calendar can be customized using the `className` prop which accepts a string of Tailwind classes.
 * Each class should follow the format `[&_.date-picker\\_\\_{element}]:{tailwind-classes}`.
 *
 * Base classes available for customization:
 * - .base-picker__input: Input field container
 * - .base-picker__popover: Popover container
 * - .date-picker__calendar: Calendar container
 * - .date-picker__button-next: Next month button
 * - .date-picker__button-previous: Previous month button
 * - .date-picker__selected: Selected dates
 * - .date-picker__today: Today's date
 * - .date-picker__weekday: Weekday headers
 * - .date-picker__day-button: Day buttons
 * - .date-picker__day: Day cells
 * - .date-picker__outside: Days outside current month
 * - .date-picker__month-caption: Month/year caption
 * - .date-picker__months: Months container
 * - .date-picker__month: Month container
 * - .date-picker__week: Week container
 * - .date-picker__range-start: Range start date
 * - .date-picker__range-end: Range end date
 * - .date-picker__range-middle: Dates in range
 * - .date-picker__nav: Navigation container
 * - .date-picker__month-grid: Month grid
 * - .date-picker__caption: Caption container
 * - .date-picker__caption-label: Caption label
 * - .date-picker__weekdays: Weekdays container
 * - .date-picker__hidden: Hidden elements
 * - .date-picker__year-grid--view: Year grid view
 * - .date-picker__month-grid--view: Month grid view
 * - .date-picker__date-footer: This is class for button footer in month and year view

**
 * # Example Usage
 *
 * When using Tailwind CSS with custom class names, we need to use `String.raw` because Tailwind
 * interprets underscores (`_`) as class separators. Without `String.raw`, the escape characters don't work
 * properly and you would need to use four backslashes (`\\\\`) to make it work. Using `String.raw` ensures
 * that the class names with underscores are properly escaped and interpreted by Tailwind.
 *
 * Here's an example of how to use the DatePicker component with custom class for styling:
 *
 * ```tsx
 * import { DatePicker } from './date-picker'
 *
 * function MyCustomDatePicker() {
 *   return (
 *     <DatePicker
 *       name="custom-date"
 *       label="Select Date"
 *       placeholder="YYYY/MM/DD"
 *       className={String.raw`
 *         [&_.date-picker\\_\\_button-next]:bg-blue-500
 *         [&_.date-picker\\_\\_button-next]:text-white
 *         [&_.date-picker\\_\\_button-next]:rounded-full
 *         [&_.date-picker\\_\\_selected]:!bg-blue-500
 *         [&_.date-picker\\_\\_selected]:!text-white
 *         [&_.date-picker\\_\\_today]:border-2
 *         [&_.date-picker\\_\\_today]:border-blue-500
 *         [&_.date-picker\\_\\_weekday]:text-blue-500
 *         [&_.date-picker\\_\\_day-button]:hover:bg-blue-100
 *       `}
 *     />
 *   )
 * }
 * ```
 *
 * > **Note:** This component does not have built-in validation. If you need built-in validation
 * > you can use the [Date Picker Field](?path=/docs/scalars-date-picker-field--readme)
 * > component.
 */
const meta: Meta<typeof DatePicker> = {
  title: 'Data Entry/Date Picker',
  component: DatePicker,
  parameters: {
    layout: 'centered',
    form: {
      resetBehavior: 'unmount',
    },
  },
  decorators: [withTimestampsAsISOStrings],
  tags: ['autodocs'],
  argTypes: {
    ...getDefaultArgTypes({
      valueControlType: 'date',
      valueType: 'date',
    }),
    minDate: {
      control: 'date',
      description: 'Minimum selectable date in the date picker',
      table: {
        type: { summary: 'string' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    maxDate: {
      control: 'date',
      description: 'Maximum selectable date in the date picker',
      table: {
        type: { summary: 'string' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    disablePastDates: {
      control: 'boolean',
      description: 'Disable past dates in the date picker',
      table: {
        type: { summary: 'boolean' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    disableFutureDates: {
      control: 'boolean',
      description: 'Disable future dates in the date picker',
      table: {
        type: { summary: 'boolean' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    dateFormat: {
      control: {
        type: 'select',
      },
      description: 'The format of the date in the date picker',
      options: Object.keys(FORMAT_MAPPING),
      table: {
        defaultValue: { summary: 'YYYY-MM-DD' },
        type: {
          summary: 'string',
        },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    weekStart: {
      control: 'select',
      options: ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'],
      table: {
        defaultValue: { summary: 'Monday' },
        type: {
          summary: 'string',
        },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    autoClose: {
      control: 'boolean',
      description: 'Close the date picker when a date is selected',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'false' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    placeholder: {
      control: 'text',
      description: 'The placeholder text for the date picker',
      table: {
        type: { summary: 'string' },
        category: StorybookControlCategory.DEFAULT,
      },
    },
    className: {
      control: 'text',
      description: 'Additional className for customizing the calendar',
      table: {
        type: { summary: 'string' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
  },
  args: {
    name: 'date-picker-field',
  },
}

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
  args: {
    name: 'date',
    label: 'Pick a date',
    placeholder: '2025/01/27',
  },
}

export const Disabled: Story = {
  args: {
    name: 'date',
    label: 'Pick a date',
    placeholder: '2025/01/27',
    disabled: true,
  },
}

export const Filled: Story = {
  args: {
    name: 'date',
    label: 'Pick a date',
    value: '2025/01/27',
  },
}

export const WithCustomizedCalendar: Story = {
  args: {
    name: 'date',
    label: 'Customized Calendar',
    placeholder: '2025/01/27',
    className: String.raw`
      [&_.date-picker\\_\\_button-next]:bg-transparent
      [&_.date-picker\\_\\_button-next]:border-2
      [&_.date-picker\\_\\_button-next]:border-cyan-400
      [&_.date-picker\\_\\_button-next]:text-cyan-400
      [&_.date-picker\\_\\_button-next]:hover:bg-cyan-400/10
      [&_.date-picker\\_\\_button-next]:hover:shadow-[0_0_15px_rgba(34,211,238,0.6)]
      [&_.date-picker\\_\\_button-next]:rounded-full
      [&_.date-picker\\_\\_button-next]:p-2
      [&_.date-picker\\_\\_button-next]:transition-all
      [&_.date-picker\\_\\_button-next]:duration-300

      [&_.date-picker\\_\\_button-previous]:bg-transparent
      [&_.date-picker\\_\\_button-previous]:border-2
      [&_.date-picker\\_\\_button-previous]:border-cyan-400
      [&_.date-picker\\_\\_button-previous]:text-cyan-400
      [&_.date-picker\\_\\_button-previous]:hover:bg-cyan-400/10
      [&_.date-picker\\_\\_button-previous]:hover:shadow-[0_0_15px_rgba(34,211,238,0.6)]
      [&_.date-picker\\_\\_button-previous]:rounded-full
      [&_.date-picker\\_\\_button-previous]:p-2
      [&_.date-picker\\_\\_button-previous]:transition-all
      [&_.date-picker\\_\\_button-previous]:duration-300

      [&_.date-picker\\_\\_selected]:!bg-cyan-400
      [&_.date-picker\\_\\_selected]:!text-black
      [&_.date-picker\\_\\_selected]:!font-mono
      [&_.date-picker\\_\\_selected]:!rounded-full
      [&_.date-picker\\_\\_selected]:hover:!bg-cyan-500
      [&_.date-picker\\_\\_selected]:transition-all
      [&_.date-picker\\_\\_selected]:duration-300

      [&_.date-picker\\_\\_today]:border-2
      [&_.date-picker\\_\\_today]:border-pink-400
      [&_.date-picker\\_\\_today]:rounded-full
      [&_.date-picker\\_\\_today]:text-pink-400

      [&_.date-picker\\_\\_weekday]:text-cyan-400
      [&_.date-picker\\_\\_weekday]:font-mono
      [&_.date-picker\\_\\_weekday]:text-sm

      [&_.date-picker\\_\\_day-button]:text-cyan-400
      [&_.date-picker\\_\\_day-button]:hover:text-cyan-300
      [&_.date-picker\\_\\_day-button]:hover:bg-cyan-400/10
      [&_.date-picker\\_\\_day-button]:rounded-full
      [&_.date-picker\\_\\_day-button]:transition-all
      [&_.date-picker\\_\\_day-button]:duration-300

      [&_.date-picker\\_\\_outside]:opacity-40
      [&_.date-picker\\_\\_outside]:!bg-gray-900

      [&_.date-picker\\_\\_month-caption]:text-cyan-400
      [&_.date-picker\\_\\_month-caption]:font-mono
      [&_.date-picker\\_\\_month-caption]:p-2
      [&_.date-picker\\_\\_month-caption]:text-lg
      [&_.date-picker\\_\\_month-caption]:border-b
      [&_.date-picker\\_\\_month-caption]:border-cyan-400/30
      [&_.date-picker\\_\\_year-grid--view]:border-2
      [&_.date-picker\\_\\_year-grid--view]:border-red-400
      [&_.date-picker\\_\\_year-grid--view]:rounded-lg
      [&_.date-picker\\_\\_year-grid--view]:p-2
      [&_.date-picker\\_\\_year-grid--view]:bg-blue-500
      [&_.date-picker\\_\\_year-grid--view]:text-white
      [&_.date-picker\\_\\_year-grid--view]:font-bold
      [&_.date-picker\\_\\_month-grid--view]:border-2
      [&_.date-picker\\_\\_month-grid--view]:border-blue-400
      [&_.date-picker\\_\\_month-grid--view]:rounded-lg
      [&_.date-picker\\_\\_month-grid--view]:p-2
      [&_.date-picker\\_\\_month-grid--view]:bg-red-500
      [&_.date-picker\\_\\_month-grid--view]:text-white
      [&_.date-picker\\_\\_month-grid--view]:font-bold
      [&_.date-picker\\_\\_month-grid--view]:button:text-blue-400
      [&_.date-picker\\_\\_date-footer]:gap-2
      [&_.date-picker\\_\\_date-footer]:[&>button]:text-blue-400
      [&_.date-picker\\_\\_date-footer]:[&>button]:hover:text-blue-500
      [&_.date-picker\\_\\_date-footer]:[&>button]:hover:bg-blue-400/10
      [&_.date-picker\\_\\_date-footer]:[&>button]:rounded-full
      [&_.date-picker\\_\\_date-footer]:[&>button]:p-2
      [&_.date-picker\\_\\_date-footer]:[&>button]:transition-all
      [&_.date-picker\\_\\_date-footer]:[&>button]:duration-300
      [&.input-field]:border-cyan-400
      [&.input-field]:rounded-md
  
    `,
  },
}
