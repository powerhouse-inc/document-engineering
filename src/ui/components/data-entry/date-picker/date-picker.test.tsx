import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { DatePicker } from './date-picker'
import { addDays, format } from 'date-fns'

// Helper function to get a date relative to today
export const getRelativeDate = (daysFromToday: number) => {
  const date = addDays(new Date(), daysFromToday)
  return format(date, 'yyyy-MM-dd')
}

const fixedDate = new Date('2025-01-01T12:00:00Z')

beforeEach(() => {
  vi.setSystemTime(fixedDate)
})

afterEach(() => {
  vi.useRealTimers()
})

describe('DatePicker', () => {
  const defaultProps = {
    name: 'test-date',
    label: 'Test Date',
  }
  it('should mock the system date to fixedDate', () => {
    const now = new Date()
    expect(now.toISOString()).toBe(fixedDate.toISOString())
  })

  it('should match the snapshot', () => {
    const { container } = render(
      <DatePicker label="Test Label" name="test-date" value="2025-01-01" dateFormat="yyyy-MM-dd" />
    )
    expect(container).toMatchSnapshot()
  })

  it('should display the label when provided', () => {
    const labelText = 'Test Label'
    render(<DatePicker name="test-date" label={labelText} />)
    expect(screen.getByText(labelText)).toBeInTheDocument()
  })

  it('should render with label', () => {
    render(<DatePicker {...defaultProps} />)
    expect(screen.getByText('Test Date')).toBeInTheDocument()
    expect(screen.getByRole('textbox')).toBeInTheDocument()
  })

  it('should show description when provided', () => {
    const description = 'This is a test description'
    render(<DatePicker {...defaultProps} description={description} />)
    expect(screen.getByText(description)).toBeInTheDocument()
  })

  it('should show error messages when provided', () => {
    const errors = ['Invalid date format']
    render(<DatePicker {...defaultProps} errors={errors} />)
    expect(screen.getByText('Invalid date format')).toBeInTheDocument()
  })

  it('should show warning messages when provided', () => {
    const warnings = ['Date is in the past']
    render(<DatePicker {...defaultProps} warnings={warnings} />)
    expect(screen.getByText('Date is in the past')).toBeInTheDocument()
  })

  it('should handle disabled state', () => {
    render(<DatePicker {...defaultProps} disabled />)
    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()
  })

  it('should disable dates before minDate', async () => {
    // Arrange: Get dates relative to today
    const minDate = getRelativeDate(5) // 5 days from today
    render(<DatePicker label="Test Label" name="test-date" minDate={minDate} />)

    // Act: Open calendar
    const calendarTrigger = screen.getByTestId('icon-fallback')
    await userEvent.click(calendarTrigger)

    // Assert: Calendar is visible
    const calendar = await screen.findByRole('dialog')
    expect(calendar).toBeInTheDocument()

    // Assert: Date before minDate is disabled
    const dateButton = screen.getByRole('button', {
      name: format(addDays(new Date(), 2), 'EEEE, MMMM do, yyyy'),
    })
    expect(dateButton).toHaveClass('disabled:pointer-events-none')
  })

  it('should disable dates after maxDate', async () => {
    // Arrange: Get dates relative to today
    const maxDate = getRelativeDate(5) // 5 days from today

    render(<DatePicker label="Test Label" name="test-date" maxDate={maxDate} />)

    // Act: Open calendar
    const calendarTrigger = screen.getByTestId('icon-fallback')
    await userEvent.click(calendarTrigger)

    // Assert: Calendar is visible
    const calendar = await screen.findByRole('dialog')
    expect(calendar).toBeInTheDocument()

    // Assert: Date after maxDate is disabled
    const dateButton = screen.getByRole('button', {
      name: format(addDays(new Date(), 7), 'EEEE, MMMM do, yyyy'),
    })
    expect(dateButton).toHaveClass('disabled:pointer-events-none')
    const input = screen.getByRole('textbox')
    expect(input).toHaveValue('')
  })

  it('should handle custom date format', async () => {
    // Arrange: Render component with custom date format
    const dateFormat = 'DD/MM/YYYY'

    render(<DatePicker {...defaultProps} dateFormat={dateFormat} />)

    // Act: Open calendar and select date
    const calendarTrigger = screen.getByTestId('icon-fallback')
    await userEvent.click(calendarTrigger)
    const dateButton = screen.getByRole('button', {
      name: format(addDays(new Date(), 3), 'EEEE, MMMM do, yyyy'),
    })
    await userEvent.click(dateButton)

    // Assert: Input value matches custom format
    const input: HTMLInputElement = screen.getByRole('textbox')
    await waitFor(() => {
      expect(input.value).toMatch(/^\d{2}\/\d{2}\/\d{4}$/)
    })
  })

  it('should close calendar when clicking outside', async () => {
    // Arrange: Render component
    render(<DatePicker {...defaultProps} />)

    // Act: Open calendar and click outside
    const calendarTrigger = screen.getByTestId('icon-fallback')
    await userEvent.click(calendarTrigger)
    await userEvent.click(document.body)

    // Assert: Calendar is closed
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument()
  })

  it('should handle manual date input', async () => {
    // Arrange: Render component
    render(<DatePicker {...defaultProps} />)
    const input: HTMLInputElement = screen.getByRole('textbox')

    // Act: Type date manually
    const testDate = getRelativeDate(3)
    await userEvent.type(input, testDate)

    // Assert: Input has correct value
    expect(input.value).toBe(testDate)
  })

  it('should open calendar when clicking the input', async () => {
    // Arrange: Render component
    render(<DatePicker {...defaultProps} />)

    // Act: Click calendar button
    const calendarButton = screen.getByTestId('icon-fallback').closest('button')
    await userEvent.click(calendarButton!)

    // Assert: Calendar is visible with grid
    const calendar = await screen.findByRole('dialog')
    expect(calendar).toBeInTheDocument()
    expect(screen.getByRole('grid')).toBeInTheDocument()

    // Act: Click outside
    await userEvent.click(document.body)

    // Assert: Calendar is closed
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument()
  })

  it('should handle autoClose prop', async () => {
    // Arrange: Render component with autoClose
    render(<DatePicker {...defaultProps} autoClose />)

    // Act: Open calendar and select date
    const calendarButton = screen.getByTestId('icon-fallback').closest('button')
    await userEvent.click(calendarButton!)

    const newDate = new Date(fixedDate)
    // We Get next day of the mockDate (fixedDate) that will always be in the calendar
    newDate.setDate(newDate.getDate() + 1)
    // We format the date to the expected format
    const dateButton = screen.getByRole('button', {
      name: format(newDate, 'EEEE, MMMM do, yyyy'),
    })
    await userEvent.click(dateButton)

    // Assert: Calendar closes automatically
    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument()
    })
  })

  it('should handle weekStart prop', async () => {
    // Arrange: Render component with weekStart
    render(<DatePicker {...defaultProps} weekStart="Monday" />)

    // Act: Open calendar
    const calendarTrigger = screen.getByTestId('icon-fallback')
    await userEvent.click(calendarTrigger)

    // Assert: Weekday headers are in correct order
    const weekdayHeaders = screen.getAllByRole('columnheader', { hidden: true })
    const weekdays = weekdayHeaders.map((header) => header.textContent)
    expect(weekdays).toEqual(['Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa', 'Su'])
  })
  describe('DateInputDiff differences', () => {
    it('should render DateInputDiff when viewMode is addition', () => {
      render(
        <DatePicker
          {...defaultProps}
          viewMode="addition"
          value={new Date('2025-10-11T12:00:00.000Z')}
          baseValue={new Date('2024-10-22T12:00:00.000Z')}
          dateFormat="YYYY-MM-DD"
        />
      )
      expect(screen.getByTestId('date-picker-diff')).toBeInTheDocument()
      expect(screen.getByTestId('icon-fallback')).toBeInTheDocument()
      expect(screen.getByText('2025-10-11')).toBeInTheDocument()
      expect(screen.queryByText('2024-10-22')).not.toBeInTheDocument()
    })

    it('should render DateInputDiff when viewMode is removal', () => {
      render(
        <DatePicker
          {...defaultProps}
          viewMode="removal"
          value={new Date('2025-02-11T12:00:00.000Z')}
          baseValue={new Date('2024-10-22T12:00:00.000Z')}
          dateFormat="YYYY-MM-DD"
        />
      )
      expect(screen.getByTestId('date-picker-diff')).toBeInTheDocument()
      expect(screen.getByTestId('icon-fallback')).toBeInTheDocument()
      expect(screen.queryByText('2025-02-11')).not.toBeInTheDocument()
      expect(screen.getByText('2024-10-22')).toBeInTheDocument()
    })

    it('should render DateInputDiff when viewMode is mixed', () => {
      render(
        <DatePicker
          {...defaultProps}
          viewMode="mixed"
          value={new Date('2025-01-01T12:00:00.000Z')}
          baseValue={new Date('2024-01-01T12:00:00.000Z')}
        />
      )
      expect(screen.getByTestId('date-picker-diff')).toBeInTheDocument()
      const iconFallbacks = screen.getAllByTestId('icon-fallback')
      expect(iconFallbacks).toHaveLength(2)
      expect(screen.getByText('2024-01-01')).toBeInTheDocument()
      expect(screen.getByText('2025-01-01')).toBeInTheDocument()
    })

    it('should not render DateInputDiff when viewMode is edition', () => {
      render(<DatePicker {...defaultProps} viewMode="edition" value="2025-01-01" />)
      expect(screen.queryByTestId('date-picker-diff')).not.toBeInTheDocument()
    })
  })
  it('should show year only years when format is YYYY', async () => {
    const date = new Date()
    const year = date.getFullYear().toString()
    render(<DatePicker {...defaultProps} dateFormat="YYYY" />)

    const calendarTrigger = screen.getByTestId('icon-fallback')
    await userEvent.click(calendarTrigger)

    expect(screen.getAllByText(year)).toHaveLength(2)
    expect(screen.queryByText('January')).not.toBeInTheDocument()
  })
  it('should show year and month only when format is YYYY-MM', async () => {
    const date = new Date()
    const year = date.getFullYear().toString()
    const month = date.toLocaleString('en-US', { month: 'long' })

    render(<DatePicker {...defaultProps} dateFormat="YYYY-MM" />)

    const calendarTrigger = screen.getByTestId('icon-fallback')
    await userEvent.click(calendarTrigger)

    expect(screen.getAllByText(year)).toHaveLength(2)
    expect(screen.getByText(month)).toBeInTheDocument()
  })
})
