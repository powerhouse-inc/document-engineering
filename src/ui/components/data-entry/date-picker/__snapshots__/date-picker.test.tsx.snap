// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DatePicker > should match the snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
    >
      Test Label
    </label>
    <div
      class="flex flex-col space-y-2"
    >
      <div
        class="flex w-full rounded-md text-sm focus-within:ring-ring dark:ring-charcoal-300 ring-gray-900 focus-within:ring-1 focus-within:ring-offset-0 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white focus:[&_.input-field]:bg-transparent dark:focus-within:hover:bg-charcoal-900 focus-within:hover:cursor-default focus-within:hover:bg-white [&.base-picker\\\\_\\\\_popover]:px-4 [&.base-picker\\\\_\\\\_popover]:pb-6 [&.base-picker\\\\_\\\\_popover]:pt-3 undefined"
      >
        <button
          aria-controls="radix-:r0:"
          aria-expanded="false"
          aria-haspopup="dialog"
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground rounded-l-md border-none pl-3 pr-2 focus:text-selected-foreground focus:bg-none button-ghost"
          data-state="closed"
          type="button"
        >
          <svg
            class="hover:none text-gray-700 dark:text-gray-50"
            data-testid="icon-fallback"
            fill="none"
            height="16"
            style="width: 16px; height: 16px;"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              id="Calendar"
            >
              <path
                clip-rule="evenodd"
                d="M4.79998 1.06665C5.09454 1.06665 5.33332 1.30543 5.33332 1.59998V2.13332H10.6667V1.59998C10.6667 1.30543 10.9054 1.06665 11.2 1.06665C11.4945 1.06665 11.7333 1.30543 11.7333 1.59998V2.13332H13.3333C14.217 2.13332 14.9333 2.84966 14.9333 3.73332V13.3333C14.9333 14.217 14.217 14.9333 13.3333 14.9333H2.66665C1.783 14.9333 1.06665 14.217 1.06665 13.3333V3.73332C1.06665 2.84966 1.78299 2.13332 2.66665 2.13332H4.26665V1.59998C4.26665 1.30543 4.50543 1.06665 4.79998 1.06665ZM10.6667 3.19998V3.73332C10.6667 4.02787 10.9054 4.26665 11.2 4.26665C11.4945 4.26665 11.7333 4.02787 11.7333 3.73332V3.19998H13.3333C13.6279 3.19998 13.8667 3.43877 13.8667 3.73332V5.33332H2.13332V3.73332C2.13332 3.43877 2.3721 3.19998 2.66665 3.19998H4.26665V3.73332C4.26665 4.02787 4.50543 4.26665 4.79998 4.26665C5.09454 4.26665 5.33332 4.02787 5.33332 3.73332V3.19998H10.6667ZM2.13332 6.39998V13.3333C2.13332 13.6279 2.3721 13.8667 2.66665 13.8667H13.3333C13.6279 13.8667 13.8667 13.6279 13.8667 13.3333V6.39998H2.13332ZM7.46665 7.99998C7.46665 7.70543 7.70543 7.46665 7.99999 7.46665C8.29454 7.46665 8.53332 7.70543 8.53332 7.99998C8.53332 8.29454 8.29454 8.53332 7.99999 8.53332C7.70543 8.53332 7.46665 8.29454 7.46665 7.99998ZM10.1333 7.46665C9.83877 7.46665 9.59999 7.70543 9.59999 7.99998C9.59999 8.29454 9.83877 8.53332 10.1333 8.53332C10.4279 8.53332 10.6667 8.29454 10.6667 7.99998C10.6667 7.70543 10.4279 7.46665 10.1333 7.46665ZM11.7333 7.99998C11.7333 7.70543 11.9721 7.46665 12.2667 7.46665C12.5612 7.46665 12.8 7.70543 12.8 7.99998C12.8 8.29454 12.5612 8.53332 12.2667 8.53332C11.9721 8.53332 11.7333 8.29454 11.7333 7.99998ZM12.2667 9.59998C11.9721 9.59998 11.7333 9.83877 11.7333 10.1333C11.7333 10.4279 11.9721 10.6667 12.2667 10.6667C12.5612 10.6667 12.8 10.4279 12.8 10.1333C12.8 9.83877 12.5612 9.59998 12.2667 9.59998ZM9.59999 10.1333C9.59999 9.83877 9.83877 9.59998 10.1333 9.59998C10.4279 9.59998 10.6667 9.83877 10.6667 10.1333C10.6667 10.4279 10.4279 10.6667 10.1333 10.6667C9.83877 10.6667 9.59999 10.4279 9.59999 10.1333ZM7.99999 9.59998C7.70543 9.59998 7.46665 9.83877 7.46665 10.1333C7.46665 10.4279 7.70543 10.6667 7.99999 10.6667C8.29454 10.6667 8.53332 10.4279 8.53332 10.1333C8.53332 9.83877 8.29454 9.59998 7.99999 9.59998ZM5.33332 10.1333C5.33332 9.83877 5.5721 9.59998 5.86665 9.59998C6.1612 9.59998 6.39998 9.83877 6.39998 10.1333C6.39998 10.4279 6.1612 10.6667 5.86665 10.6667C5.5721 10.6667 5.33332 10.4279 5.33332 10.1333ZM3.73332 9.59998C3.43877 9.59998 3.19999 9.83877 3.19999 10.1333C3.19999 10.4279 3.43877 10.6667 3.73332 10.6667C4.02787 10.6667 4.26665 10.4279 4.26665 10.1333C4.26665 9.83877 4.02787 9.59998 3.73332 9.59998ZM3.19999 12.2667C3.19999 11.9721 3.43877 11.7333 3.73332 11.7333C4.02787 11.7333 4.26665 11.9721 4.26665 12.2667C4.26665 12.5612 4.02787 12.8 3.73332 12.8C3.43877 12.8 3.19999 12.5612 3.19999 12.2667ZM5.86665 11.7333C5.5721 11.7333 5.33332 11.9721 5.33332 12.2667C5.33332 12.5612 5.5721 12.8 5.86665 12.8C6.1612 12.8 6.39998 12.5612 6.39998 12.2667C6.39998 11.9721 6.1612 11.7333 5.86665 11.7333ZM7.46665 12.2667C7.46665 11.9721 7.70543 11.7333 7.99999 11.7333C8.29454 11.7333 8.53332 11.9721 8.53332 12.2667C8.53332 12.5612 8.29454 12.8 7.99999 12.8C7.70543 12.8 7.46665 12.5612 7.46665 12.2667ZM10.1333 11.7333C9.83877 11.7333 9.59999 11.9721 9.59999 12.2667C9.59999 12.5612 9.83877 12.8 10.1333 12.8C10.4279 12.8 10.6667 12.5612 10.6667 12.2667C10.6667 11.9721 10.4279 11.7333 10.1333 11.7333Z"
                fill="currentcolor"
                fill-rule="evenodd"
                id="Shape"
              />
            </g>
          </svg>
        </button>
        <input
          class="flex h-9 rounded-md text-sm font-normal leading-5 text-gray-900 dark:text-gray-50 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white px-3 py-2 font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600 focus-visible:outline-none focus-visible:ring-gray-900 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 disabled:cursor-not-allowed disabled:border-gray-300 disabled:bg-gray-50 disabled:text-gray-700 disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300 w-full rounded-l-none border-none text-right placeholder:text-right input-field focus:bg-white focus-visible:ring-0 focus-visible:ring-offset-0"
          data-cast="DateString:yyyy-MM-dd"
          label="Test Label"
          name="test-date"
          value="2025-01-01"
        />
      </div>
    </div>
  </div>
</div>
`;
